### PR TITLE
er-aml implementation

### PR DESCRIPTION
**Pull request by GeremiaPompei**
Implementation of er-aml criterion of the paper: https://openreview.net/forum?id=N8MaByOzUfb

### CODE REVIEW

**Review by AntonioCarta — CHANGES_REQUESTED** [review]

**Comment by AlbinSou** [comment]
> > 
> 
> paper results:
> 
> * cifar10: 0.52 accuracy
> * cifar100: 0.24 accuracy
> 
> continual-learning-baselines results:
> 
> * cifar10: 0.49 accuracy
> * cifar100: 0.25 accuracy

Hey @GeremiaPompei, thank you for the implementation. I left a few comments on the code. Most importantly I think there is the fix in the way the memory buffer is sampled and the possibility to fetch negatives from only current classes instead of all classes. I think one good way to check if the implementation is working as in the paper would be to compare the results of ER-ACE to the results of ER-AML on some benchmarks where they conclude that AML works better. If we get similar or slightly higher results on these benchmarks then it would good news.

**Comment by AntonioCarta** [comment]
> Hey @GeremiaPompei, thank you for the implementation. I left a few comments on the code

Did you publish the review? because I can't see the comments.

**Comment by GeremiaPompei** [comment]
Hi @AlbinSou, thank you for the suggestions. In general I followed the original repository of the paper https://github.com/pclucas14/AML for the implementation. According to sampling negatives I can't see the comments; as soon as I will see them I change the implementation. About comparison between ER_ACE and ER-AML seems that they reach similar resulta. I have already implemented the tests of er-aml in https://github.com/ContinualAI/continual-learning-baselines repository where are used the same benchmarks.

**Review by AlbinSou — CHANGES_REQUESTED** [review]

File: avalanche/training/supervised/er_aml.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                self.loss += self.criterion()
+            else:
+                buffer_replay_data = next(
+                    cycle(
+                        torch.utils.data.DataLoader(
+                            self.storage_policy.buffer,
+                            batch_size=len(self.storage_policy.buffer),
+                        )
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (GeremiaPompei) [reply avalanche/training/supervised/er_aml.py:146]: I agree, at this point maybe it's better to assign the pos neg sample function (__sample_pos_neg in AMLCriterion) to the buffer replay (extending for example ClassBalancedBuffer with a subclass that add the sample_pos_neg method). In this way it is able to sample directly from the total memory avoiding to use the data loader

File: avalanche/training/regularization.py [hunk 2]
```diff
@@ -166,4 +166,157 @@ def __call__(self, out_in, target_in, out_buffer, target_buffer):
...
+        """
+        x_all = torch.cat((x_buffer, x_in))
+        y_all = torch.cat((y_buffer, y_in))
+        indexes = torch.arange(y_all.shape[0]).to(self.device)
+
+        same_x = indexes[-x_in.shape[0] :].reshape(1, -1) == indexes.reshape(-1, 1)
+        same_y = y_in.reshape(1, -1) == y_all.reshape(-1, 1)
+
+        valid_pos = same_y & ~same_x
+        valid_neg = ~same_y
```
[the remark this answers is not in the document]
Comment (GeremiaPompei) [reply avalanche/training/regularization.py:220]: Ok, I will add this option

File: avalanche/training/regularization.py [hunk 3]
```diff
@@ -166,4 +166,157 @@ def __call__(self, out_in, target_in, out_buffer, target_buffer):
...
+        @param target_in: Labels of new examples.
+        @param output_buffer: Predictions of samples from buffer.
+        @param target_buffer: Labels of samples from buffer.
+        @param buffer_replay_data: Buffer replay data to compute positive and negative samples.
+        @return: ER_AML computed loss.
+        """
+        x_buffer, y_buffer, _ = buffer_replay_data
+        pos_x, pos_y, neg_x, neg_y, is_invalid = self.__sample_pos_neg(
+            input_in, target_in, x_buffer, y_buffer
+        )
```
[the remark this answers is not in the document]
Comment (GeremiaPompei) [reply avalanche/training/regularization.py:298]: I tried to use a batch instead of the total memory for sampling positives and negatives and results are:
- cifar10: 0.48
- cifar100: 0.21

instead with the whole memory results are:
- cifar10: 0.49 accuracy
- cifar100: 0.25 accuracy

**Comment by GeremiaPompei** [comment]
@AntonioCarta and @AlbinSou I update the code 
- adding the option for removing negative sampling of other tasks. This is possible using a flag as an hyperparam of ER_ML strategy. Default value is true as the original paper
- optimizing the pos neg sampling extracting each time a number of examples for each class in the replay_loader equal to the number of new examples. Now the algorithm seem to scale better

let me know if there is something wrong or not so clear that I will modify it

**Comment by AlbinSou** [comment]
@GeremiaPompei it looks alright to me. Have the results changed after these changes ?

**Comment by GeremiaPompei** [comment]
@AlbinSou results are

paper results:
cifar10: 0.52 accuracy
cifar100: 0.24 accuracy

continual-learning-baselines results:
cifar10: 0.47 accuracy
cifar100: 0.25 accuracy
